tests/integration/utils.py

```diff
@@ -0,0 +1,124 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+import hashlib
+import json
+import logging
+import os
+from pathlib import Path
+from subprocess import PIPE, Popen
+from typing import Dict, List, Tuple
+
+from git import Repo
+
+log = logging.getLogger(__name__)
+
+
+class TestParameters:
+    def __init__(
+        self,
+        *,
+        repo: str = "",
+        ref: str = "",
+        test_case: str = "",
+        packages: Tuple = (),
+        check_files: List = [],
+        expected_rc: int = None,
+        expected_output: str = None,
+    ):
+        """Construct test case arguments."""
+        self.repo = repo
+        self.ref = ref
+        self.test_case = test_case
+        self.packages = packages
+        self.check_files = check_files
+        self.expected_rc = expected_rc
+        self.expected_output = expected_output
+
+
+class ContainerImage:
+    def __init__(self, repository: str):
+        """Initialize ContainerImage object with associated repository."""
+        self.repository = repository
+
+    def pull_image(self):
+        cmd = ["podman", "pull", self.repository]
+        run_cmd(cmd)
+        log.info("Pulled image: %s.", self.repository)
+
+    def run_cmd_on_image(self, cmd: List, tmpdir: Path):
+        image_cmd = ["podman", "run", "-v", f"{tmpdir}:{tmpdir}:z", self.repository] + cmd
+        return run_cmd(image_cmd)
+
+
+def clone_repository(repo: str, ref: str, folder_name: str, tmpdir: Path) -> Path:
+    """
+    Clone repository and checkout specific commit.
+
+    :param str repo: Git repository URL
+    :param str ref: Git reference
+    :param str folder_name: Name of folder where content will be cloned
+    :param str tmpdir: Temp directory for pytest
+    :return: Absolute path to cloned repository
+    :rtype: str
+    """
+    folder = os.path.join(tmpdir, folder_name)
+
+    repo = Repo.clone_from(repo, folder)
+    repo.git.checkout(ref)
+    log.info("Cloned repository path: %s", folder)
+    return folder
```

Comment:
I think this is a string, not a Path as the type annotation says.

You could make it a Path with
```python
folder = tmpdir / folder_name
```

I've submitted a PR to start type-checking tests: https://github.com/containerbuildsystem/cachi2/pull/42

Reply:
I was wondering why I did not catch any annotation type issues. :D Thanks for opening that PR. I'll fix those here.